filestore/fsrefstore.go

```diff
@@ -0,0 +1,190 @@
+package filestore
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/ipfs/go-ipfs/blocks"
+	"github.com/ipfs/go-ipfs/blocks/blockstore"
+	pb "github.com/ipfs/go-ipfs/filestore/pb"
+	dshelp "github.com/ipfs/go-ipfs/thirdparty/ds-help"
+	posinfo "github.com/ipfs/go-ipfs/thirdparty/posinfo"
+
+	ds "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore"
+	dsns "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore/namespace"
+	dsq "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore/query"
+	proto "gx/ipfs/QmT6n4mspWYEya864BhCUJEgyxiRfmiSY9ruQwTUNpRKaM/protobuf/proto"
+	cid "gx/ipfs/QmcTcsTvfaeEBRFo1TkFgT8sRmgi1n1LTZpecfVP8fzpGD/go-cid"
+)
+
+var FilestorePrefix = ds.NewKey("filestore")
+
+type FileManager struct {
+	ds   ds.Batching
+	root string
+}
+
+type CorruptReferenceError struct {
+	Err error
+}
+
+func (c CorruptReferenceError) Error() string {
+	return c.Err.Error()
+}
+
+func NewFileManager(ds ds.Batching, root string) *FileManager {
+	return &FileManager{dsns.Wrap(ds, FilestorePrefix), root}
+}
+
+func (f *FileManager) AllKeysChan(ctx context.Context) (<-chan *cid.Cid, error) {
+	q := dsq.Query{KeysOnly: true}
+	q.Prefix = FilestorePrefix.String()
+
+	res, err := f.ds.Query(q)
+	if err != nil {
+		return nil, err
+	}
+
+	out := make(chan *cid.Cid)
+	go func() {
+		defer close(out)
+		for {
+			v, ok := res.NextSync()
+			if !ok {
+				return
+			}
+
+			k := ds.RawKey(v.Key)
+			c, err := dshelp.DsKeyToCid(k)
+			if err != nil {
+				log.Error("decoding cid from filestore: %s", err)
+				continue
+			}
+
+			select {
+			case out <- c:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return out, nil
+}
+
+func (f *FileManager) DeleteBlock(c *cid.Cid) error {
+	err := f.ds.Delete(dshelp.CidToDsKey(c))
+	if err == ds.ErrNotFound {
+		return blockstore.ErrNotFound
+	}
+	return err
+}
+
+func (f *FileManager) Get(c *cid.Cid) (blocks.Block, error) {
+	o, err := f.ds.Get(dshelp.CidToDsKey(c))
+	switch err {
+	case ds.ErrNotFound:
+		return nil, blockstore.ErrNotFound
+	default:
+		return nil, err
+	case nil:
+		//
+	}
+
+	data, ok := o.([]byte)
+	if !ok {
+		return nil, fmt.Errorf("stored filestore dataobj was not a []byte")
+	}
+
+	var dobj pb.DataObj
+	if err := proto.Unmarshal(data, &dobj); err != nil {
+		return nil, err
+	}
+
+	out, err := f.readDataObj(&dobj)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Ah, yeah. `NewBlockWithCid` only does that check if `u.Debug` is true. good catch